### ISSUE TITLE
Using Hexagon DSP on consumer hardware

### ISSUE THREAD

**Issue by mirsadm** [halide/Halide#5200]
I am wondering if it is possible use the Hexagon DSP on a consumer device? I'm interested in comparing the difference in performance between the CPU and DSP on my device but have not had much luck in getting it to work. I'm using the Halide distribution from the Hexagon SDK. It appears to support an "unsigned" mode on newer devices (Snapdragon 855). I'm getting errors ranging from unable to load 'libhalide_hexagon_host.so' to access denied when running dlopen on 'libhalide_hexagon_remote_skel.so' if I package the shared library with my app.

Any help would be appreciated.

**Comment by dsharletg** [comment]
@pranavb-ca @dpalermo would be best equipped to answer this.

My experience is that you need devices that allow unsigned or debug-signed code to run. However, if you see an "unsigned" mode, it seems like that should work....?

Other obstacles you might be encountering are:
- SElinux might be blocking the IO used to communicate with Hexagon.
- If your device does not whitelist libraries, they can't be loaded from the system paths via dlopen (this is the Android OS dlopen for libhalide_hexagon_host.so, rather than the Hexagon dlopen used for libhalide_hexagon_remote_skel.so). The best way to avoid this is to ship the libraries in your app/put them in a local folder if a shell app (this is recommended anyways), so you don't need permission from the system to load the libraries. Our 'adb_run_on_device.sh' scripts illustrate how to do this with ADSP_LIBRARY_PATH: https://github.com/halide/Halide/blob/master/apps/blur/adb_run_on_device.sh#L9

**Comment by pranavb-ca** [comment]
Indeed, on 855 there is unsigned mode execution that is available. I checked and realized we haven't upstream this yet. But, there is not reason not to. Let me push it up in a bit. Other than that like @dsharletg mentioned, if you are writing an app (apk) you need to ensure that libhalide_hexagon_host.so is packaged in your app itself (https://developer.android.com/studio/projects/gradle-external-native-builds#jniLibs). And again as mentioned before, you need to set up ADSP_LIBRARY_PATH to access libhalide_hexagon_remote_skel.so

**Comment by mirsadm** [comment]
@dsharletg @pranavb-ca Thanks for the help. I did manage to get it working by uploading the shared libraries to the SD card and setting the environment variable to point to that instead. For whatever reason packaging inside the APK didn't seem to work. They didn't seem to be available outside the APK. Right now it seems to fail to allocate any size memory with ion_malloc() when copying to the device. I haven't had much time to investigate this issue either, I could just be using it incorrectly I guess.

**Comment by Neustradamus** [comment]
Hello guys,

I wish you a Happy New Year!

I contact you because you work on Hexagon DSP SDK.

And a lot of people have problems with AC3 and the Snapdragon 835 (Qualcomm Hexagon 682 DSP):
- https://dev.freebox.fr/bugs/task/23233

I think a translator is needed.

What do you think?

Or who can help?

Thanks in advance.

**Comment by anuejn** [comment]
Here is a late but hopefully for some future person helpful comment on this:
Ideally there would be upstream support for requesting an unsigned dsp context in halide. However while that is not the implemented, this snippet can provide a workaround:
```
#include <HalideRuntime.h>
#include <stdio.h>
#include <cstdint>

typedef int (*remote_session_control_fn)(uint32_t req, void *data,
                                         uint32_t datalen);

#define DSPRPC_CONTROL_UNSIGNED_MODULE 2

struct remote_rpc_control_unsigned_module {
  int domain;
  int enable;
};

int enable_cdsp_unsigned() {
  void *lib = halide_load_library("libcdsprpc.so");
  if (lib == NULL) {
    printf("Failed to load libcdsprpc.so\n");
    return 1;
  }
  remote_session_control_fn remote_session_control =
      (remote_session_control_fn)halide_get_library_symbol(
          lib, "remote_session_control");
  if (!remote_session_control) {
    printf("Failed to find remote_session_control in libcdsprpc.so\n");
    return 1;
  }
  remote_rpc_control_unsigned_module control = {.domain = -1, .enable = 1};
  remote_session_control(DSPRPC_CONTROL_UNSIGNED_MODULE, (void *)&control,
                         sizeof(remote_rpc_control_unsigned_module));
  return 0;
}
```

Just call `enable_cdsp_unsigned` before using any dsp functionality and you should be good.
